### PR TITLE
Toolchain build improvements

### PR DESCRIPTION
**Pull request by Vogtinator**
Refactor build_toolchain.sh and update to GCC 10.1.0.

### CODE REVIEW

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 1]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
-		exit 1
+downloadAndExtract() {
+	url="$1"
+	target="download/$(basename "${url}")"
+	if ! wget -c "$url" -O "${target}"; then
+		rm -f "${target}"
+		return 1
 	fi
-fi
+	tar -xaf "${target}" --checkpoint=.10 -C download
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:72]: `--checkpoint` doesn't exist on macOS/[BSD tar](https://www.freebsd.org/cgi/man.cgi?tar(1)).
If we really want it, we could probably pipe to `pv` etc. but meh?

Comment (Vogtinator) [reply ndless-sdk/toolchain/build_toolchain.sh:72]: Is there some equivalent option?

Comment (adriweb) [reply ndless-sdk/toolchain/build_toolchain.sh:72]: Well, I see alternatives proposed here: https://superuser.com/questions/168749/is-there-a-way-to-see-any-tar-progress-per-file

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 2]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
+}
 
 # Section 1: GNU Binutils.
-echo "Building Binutils..."
-[ -f .built_binutils ] || (cd build && rm -rf ./* && ../${BINUTILS}/configure ${OPTIONS_BINUTILS} && make $PARALLEL all && make install && cd .. && rm -rf build/* && touch .built_binutils) || exit 1;
+if [ "$(cat .built_binutils 2>/dev/null)" != "${BINUTILS}" ]; then
+	if [ ! -d "download/${BINUTILS}" ]; then
+		echo "Downloading Binutils..."
+		rm -rf download/binutils*
+		downloadAndExtract http://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:80]:
```suggestion
		downloadAndExtract https://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2
```

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 3]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
 # Section 3: Newlib.
-echo "Building Newlib..."
-[ -f .built_newlib ] || (cd build && rm -rf ./* && ../${NEWLIB}/configure ${OPTIONS_NEWLIB} && make $PARALLEL && make install && cd .. && rm -rf build/* && touch .built_newlib) || exit 1;
-# Workaround for newlib bug
-rm -f "${PREFIX}/arm-none-eabi/sys-include/newlib.h"
+if [ "$(cat .built_newlib 2>/dev/null)" != "${NEWLIB}" ]; then
+	if [ ! -d "download/${NEWLIB}" ]; then
+		echo "Downloading Newlib..."
+		rm -rf download/newlib*
+		downloadAndExtract ftp://sourceware.org/pub/newlib/${NEWLIB}.tar.gz
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:116]: https seems to be working here as well

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 4]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
+fi
 
 # Section 4: GCC, step 2. Yes, this is necessary.
-echo "Building GCC (step 2)..."
-[ -f .built_gcc_step2 ] || (cd build && ../${GCC}/configure ${OPTIONS_GCC} && make $PARALLEL && make install && cd .. && rm -rf build/* && touch .built_gcc_step2) || exit 1
+if [ "$(cat .built_gcc_step2 2>/dev/null)" != "${GCC}" ]; then
+	if [ ! -d "download/${GCC}" ]; then
+		echo "Downloading GCC..."
+		rm -rf download/gcc*
+		downloadAndExtract ftp://ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:136]:
```suggestion
		downloadAndExtract https://ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz
```

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 5]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
+fi
 
 # Section 5: GDB.
-echo "Building GDB..."
-[ -f .built_gdb ] || (cd build && rm -rf ./* && ../${GDB}/configure ${OPTIONS_GDB} && make $PARALLEL && make install && cd .. && rm -rf build/* && touch .built_gdb) || exit 1;
+if [ "$(cat .built_gdb 2>/dev/null)" != "${GDB}" ]; then
+	if [ ! -d "download/${GDB}" ]; then
+		echo "Downloading GDB..."
+		rm -rf download/gdb*
+		downloadAndExtract ftp://ftp.gnu.org/gnu/gdb/${GDB}.tar.xz
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:154]:
```suggestion
		downloadAndExtract https://ftp.gnu.org/gnu/gdb/${GDB}.tar.xz
```

File: ndless-sdk/toolchain/build_toolchain.sh [hunk 6]
```diff
@@ -49,45 +56,117 @@ if ! gcc -lgmp       test.c -o test; then error=1; echo 'GMP (gmp-devel/libgmp-d
...
+fi
 
 # Section 2: GCC, step 1.
-echo "Building GCC (step 1)..."
-[ -f .built_gcc_step1 ] || (cd build && rm -rf ./* && ../${GCC}/configure ${OPTIONS_GCC} && make $PARALLEL all-gcc && make install-gcc && cd .. && rm -rf build/* && touch .built_gcc_step1) || exit 1;
+if [ "$(cat .built_gcc_step1 2>/dev/null)" != "${GCC}" ]; then
+	if [ ! -d "download/${GCC}" ]; then
+		echo "Downloading GCC..."
+		rm -rf download/gcc*
+		downloadAndExtract ftp://ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz
```
Comment (adriweb) [inline ndless-sdk/toolchain/build_toolchain.sh:98]:
```suggestion
		downloadAndExtract https://ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz
```

File: ndless-sdk/toolchain/build_toolchain.sh (hunk 1 above)
Comment (Vogtinator) [reply ndless-sdk/toolchain/build_toolchain.sh]: Let's just drop this, then. I added an echo so that it doesn't seem stuck.

**Pull request merged**